TLS Support added

jamestoyer opened the pull request:
Allows for listening to request on TLS. If the `TLSConfig` property is not set, it is assumed that no TLS listening is required

jamestoyer committed:
TLS Support added

jmcampanini commented on baseapp/server.go:
```diff
@@ -102,6 +102,13 @@ func (s *Server) Start() error {
+	tlsConfig := s.config.TLSConfig
+
+	if tlsConfig != nil {
+		return http.ListenAndServeTLS(addr, s.config.TLSConfig.CertificateFile, s.config.TLSConfig.PrivateKeyFile, s.mux)
```
does validation (existing and being valid) take place on first request or at this call? if on first request, shoudl we at least check file exists?

jamestoyer replied:
`ListenAndServeTLS` does a check in for the existence and validity of the files, returning an `error` when they are invalid. Depending on conventions, I'm happy to add validation here too

jmcampanini replied:
naw, just wanted to make sure it wouldnt wait to fail on first request. lgtm

jmcampanini replied:
one last bit: why use `s.config.TLSConfig.CertificateFile` instead of just `tlsConfig.CertificateFile`

jamestoyer replied:
Whoops, my bad. I meant to update that

jmcampanini reviewed (commented):
seems reasonable and straight forward. addition to README too?

jamestoyer committed:
TLS Config reference updated

bluekeyes commented on baseapp/server.go:
```diff
@@ -102,6 +102,12 @@ func (s *Server) Start() error {
+
+	tlsConfig := s.config.TLSConfig
+	if tlsConfig != nil {
+		return http.ListenAndServeTLS(addr, tlsConfig.CertificateFile, tlsConfig.PrivateKeyFile, s.mux)
```
In our initial internal use case, I believe we'll need to further configure some of the options that are only available via `http.Server.TLSConfig`. Instead of using the `ListenAndServe*` package functions, I think we should add an `http.Server` struct to our custom `Server` type and provide some method for people to modify it. Then the existence of our `TLSConfig` will determine which `ListenAndServe*` method is used.

bluekeyes commented on baseapp/config.go:
```diff
@@ -14,10 +14,16 @@
 
+type TLSConfig struct {
+	CertificateFile string `yaml:"cert_file" json:"certFile"`
+	PrivateKeyFile  string `yaml:"private_key_file" json:"privateKeyFile"`
```
The standard library calls these "cert file" and "key file". Should we match that terminology here? The abbreviated `cert_file` matched with the expanded `private_key_file` is bothering me.

jamestoyer replied:
I'll change them to match the standard library for consistency

jamestoyer committed:
Cert properties updated to match standard library

jamestoyer committed:
HTTP server exposed to allow for TLS overriding

jamestoyer commented on baseapp/server.go:
```diff
@@ -33,6 +34,7 @@ type Server struct {
 	middleware []func(http.Handler) http.Handler
 	logger     zerolog.Logger
 	mux        *goji.Mux
+	server     *http.Server
```
@bluekeyes I've exposed the server object here to allow for updates to it later

bluekeyes replied:
Thanks. As this is not an exported field, we'll also need a method to access it. I can think of two approaches right now:

1. Define a `Server() *http.Server` method. Simple, but might imply that calling `ListenAndServe` directly on the returned value is a good idea. Also requires callers to create the `baseapp.Server` first and then manipulated it.

2. Define a `WithServer(*http.Server)` option function that allows the user to provide their own `http.Server` instance. If not provided, we use a default and otherwise we set any values (like the address) that are unset and use it.

I like (2), as it matches how we've configured other things like the metrics registry.

jamestoyer commented on baseapp/server.go:
```diff
@@ -69,6 +71,13 @@ func NewServer(c HTTPConfig, params ...Param) (*Server, error) {
+	base.server = &http.Server{
+		Addr:      addr,
+		Handler:   base.mux,
+		TLSConfig: &tls.Config{},
```
Should I be setting this? Or should I not and leave the end consumer set it up?

bluekeyes replied:
I'd leave it unset to match the default behavior we had before.

jamestoyer committed:
Default TLS config removed

Removing the default creation of `tls.Config` better replicates the
original behaviour of go-baseapp

jamestoyer committed:
WithHTTPServer param created to generate server

Allows for a default implementation to be overriden by clients that need
a more complicated HTTP Server set up

jamestoyer committed:
HTTP Server nil checking moved to server

Allows for `New` to be called and the server to just work without
needing to call anything extra. Address and Mux are set when they are
not set
